src/unistd.rs

```diff
@@ -728,34 +728,85 @@ pub fn sethostname<S: AsRef<OsStr>>(name: S) -> Result<()> {
     Errno::result(res).map(drop)
 }
 
-/// Get the host name and store it in the provided buffer, returning a pointer
-/// the `CStr` in that buffer on success (see
+/// Get the host name, returning a `CString` on success (see
 /// [gethostname(2)](http://pubs.opengroup.org/onlinepubs/9699919799/functions/gethostname.html)).
 ///
-/// This function call attempts to get the host name for the running system and
-/// store it in a provided buffer.  The buffer will be populated with bytes up
-/// to the length of the provided slice including a NUL terminating byte.  If
-/// the hostname is longer than the length provided, no error will be provided.
-/// The posix specification does not specify whether implementations will
-/// null-terminate in this case, but the nix implementation will ensure that the
-/// buffer is null terminated in this case.
+/// This function call attempts to get the host name for the running system
+/// and return it as a NUL-terminated string.
+///
+/// # Examples
 ///
 /// ```no_run
 /// use nix::unistd;
 ///
-/// let mut buf = [0u8; 64];
-/// let hostname_cstr = unistd::gethostname(&mut buf).expect("Failed getting hostname");
-/// let hostname = hostname_cstr.to_str().expect("Hostname wasn't valid UTF-8");
+/// let hostname_cstr = unistd::gethostname().expect("Failed getting hostname");
+/// let hostname = hostname_cstr.into_string().expect("Hostname wasn't valid UTF-8");
 /// println!("Hostname: {}", hostname);
 /// ```
-pub fn gethostname(buffer: &mut [u8]) -> Result<&CStr> {
-    let ptr = buffer.as_mut_ptr() as *mut c_char;
-    let len = buffer.len() as size_t;
+pub fn gethostname() -> Result<CString> {
+    // Hostname maximum length as defined by POSIX,
+    // http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/limits.h.html
+    const POSIX_HOST_NAME_MAX: usize = 255;
+    let buf = vec![0; POSIX_HOST_NAME_MAX + 1];
+    let ptr = unsafe { CString::from_vec_unchecked(buf).into_raw() };
+
+    let res = unsafe { libc::gethostname(ptr, POSIX_HOST_NAME_MAX + 1) };
+    Errno::result(res).map(|_| {
+        unsafe {
+            // Ensure returned string is always null-terminated
+            *ptr.offset(POSIX_HOST_NAME_MAX as isize) = 0;
+            CString::from_raw(ptr)
+        }
+    })
```

Comment:
@asomers Does this comment still stand or is it not relevant anymore?

Reply:
Unfortunately GitHub has forgotten the context of that comment.  But I think my objection was because he tried to handle that error condition and didn't do it correctly.  Now, he isn't handling any error conditions, which is better.